### PR TITLE
Forbid readContext() during render-phase class setState() reducer form

### PR DESCRIPTION
**Pull request by gaearon**
## Depends on https://github.com/facebook/react/pull/14672

Addresses https://github.com/facebook/react/pull/14653#issuecomment-456582655 (I think?)

Not sure we want to actually do it but figured I'd send a PR.

### CODE REVIEW

File: packages/react-reconciler/src/__tests__/ReactNewContext-test.internal.js [hunk 1]
```diff
@@ -1347,6 +1347,29 @@ describe('ReactNewContext', () => {
...
+          return null;
+        }
+      }
+
+      ReactNoop.render(<Cls />);
+      expect(() =>
+        expect(ReactNoop.flush).toThrow(
+          'Context can only be read while React is rendering',
+        ),
+      ).toWarnDev('Cannot update during an existing state transition', {
```
Comment (gaearon) [inline packages/react-reconciler/src/__tests__/ReactNewContext-test.internal.js]: We warn anyway so maybe it isn't worth it?

**Comment by acdlite** [comment]
The reason to do this is because the semantics are currently undefined — what happens when context changes? Re-rendering will have no effect because the update already happened — and we want to preserve option value to possibly support it properly in the future. Like for code splitting reducers. Same reason we warn about reading context inside Hook reducer.

`getDerivedStateFromProps` is fine because we it every time.

**Comment by gaearon** [comment]
That's true but we already unconditionally warn for render phase `setState` anyway. You can see it in the test. Usually if there's a warning, historically we didn't mind undefined behavior. It's meant to not officially be supported.

**Comment by sebmarkbage** [comment]
Added a comment that I don’t think we should do this in prod.

https://github.com/facebook/react/pull/14653#issuecomment-456859212

**Comment by gaearon** [comment]
Do we need an extra warning for this specifically? This test already warns with "Cannot update during an existing state transition".

**Comment by gaearon** [comment]
I made this depend on https://github.com/facebook/react/pull/14672 and added a similar DEV-only warning on top.

**Comment by gaearon** [comment]
It's annoying to have so many small ones. I'll send a combined one.

**Comment by gaearon** [comment]
Superseded by https://github.com/facebook/react/pull/14677

**Pull request closed without merging**